File: backend/models/product.py
from typing import Literal, Optional

from pydantic import AwareDatetime, BaseModel, Field, computed_field
from models.platform_listing import DataStatus, PlatformListing


class ProductCreate(BaseModel):
    company: str = Field(min_length=1)
    brand: str = Field(min_length=1)
    product_name: str = Field(min_length=1)
    category: str = Field(min_length=1)
    size: str = Field(min_length=1)
    barcode: str = ''
    price: Optional[float] = Field(default=None, ge=0, allow_inf_nan=False)
    currency: str = Field(pattern=r"^[A-Z]{3}$")
    image_url: Optional[str] = None
    company_id: Optional[str] = None
    brand_id: Optional[str] = None
    subcategory: Optional[str] = None
    purpose: Optional[str] = None
    variant: Optional[str] = None
    size_value: Optional[float] = Field(default=None, gt=0, allow_inf_nan=False)
    unit: Optional[Literal["L", "ml", "kg", "g", "diaper", "razor"]] = None
    pack_quantity: Optional[int] = Field(default=None, gt=0)
    attributes: dict[str, str] = Field(default_factory=dict)


class Product(ProductCreate):
    id: int = Field(gt=0)
    listings: list[PlatformListing] = Field(default_factory=list)
    source_state: DataStatus = "MANUAL"
    quality_flags: list[str] = Field(default_factory=list)
    product_family: Optional[str] = None
    verified: bool = False
    updated_at: Optional[str] = None

    @computed_field
    @property
    def data_status(self) -> DataStatus:
        observations = [item for item in self.listings if item.data_source == "verified"] or self.listings
        statuses = {item.data_status for item in observations}
        if not statuses:
            return self.source_state
        if "ERROR" in statuses:
            return "ERROR"
        if "STALE" in statuses:
            return "STALE"
        if statuses == {"DEMO"}:
            return "DEMO"
        return next(iter(statuses)) if len(statuses) == 1 else self.source_state


class PriceSummary(BaseModel):
    currency: str
    data_source: Literal["demo", "verified", "none"]
    lowest_price: Optional[float] = None
    highest_price: Optional[float] = None
    best_deals: list[PlatformListing] = Field(default_factory=list)
    number_of_platforms: int
    last_updated: Optional[AwareDatetime] = None


class ProductComparison(PriceSummary):
    product: Product
    image_url: Optional[str] = None
    listings: list[PlatformListing]

File: frontend/types/product.ts
export type DataStatus = "LIVE" | "VERIFIED" | "STALE" | "DEMO" | "ERROR" | "IMPORTED" | "MANUAL" | "COLLECTED" | "UNAVAILABLE";

export interface PlatformListing {
  platform: string;
  platform_product_id: string;
  product_name: string;
  price: number;
  original_price: number | null;
  currency: string;
  product_url: string | null;
  image_url: string | null;
  availability: "available" | "out_of_stock" | "unknown";
  last_updated: string;
  data_source: "demo" | "verified";
  data_status?: DataStatus;
  source_name?: string;
}

export interface Product {
  id: number;
  company: string;
  brand: string;
  product_name: string;
  category: string;
  size: string;
  barcode: string;
  price: number | null;
  currency: string;
  image_url?: string | null;
  listings?: PlatformListing[];
  data_status?: DataStatus;
  subcategory?: string | null;
  purpose?: string | null;
  variant?: string | null;
  size_value?: number | null;
  unit?: string | null;
  pack_quantity?: number | null;
  attributes?: Record<string, string>;
  source_state?: DataStatus;
  quality_flags?: string[];
  verified?: boolean;
  product_family?: string | null;
}

export function isProduct(value: unknown): value is Product {
  if (typeof value !== "object" || value === null) return false;
  const p = value as Record<string, unknown>;
  return Number.isInteger(p.id) && (p.price === null || (typeof p.price === "number" && Number.isFinite(p.price)))
    && ["company", "brand", "product_name", "category", "size", "barcode", "currency"].every(key => typeof p[key] === "string")
    && (p.image_url == null || typeof p.image_url === "string")
    && (p.listings === undefined || (Array.isArray(p.listings) && p.listings.every(isListing)));
}

function isListing(value: unknown): value is PlatformListing {
  if (!value || typeof value !== "object") return false;
  const item = value as Record<string, unknown>;
  return ["platform", "platform_product_id", "product_name", "currency", "last_updated"].every(key => typeof item[key] === "string")
    && typeof item.price === "number" && Number.isFinite(item.price) && item.price >= 0
    && (item.original_price === null || (typeof item.original_price === "number" && Number.isFinite(item.original_price)))
    && ["image_url", "product_url"].every(key => item[key] === null || typeof item[key] === "string")
    && ["available", "out_of_stock", "unknown"].includes(String(item.availability))
    && ["demo", "verified"].includes(String(item.data_source))
    && Number.isFinite(Date.parse(String(item.last_updated)));
}
